src/poprox_recommender/pipeline.py

```diff
@@ -71,14 +77,78 @@ def run_component(self, component_spec: ComponentSpec, state: PipelineState):
 
         output = component_spec.component(*arguments)
 
-        if not isinstance(output, (ArticleSet, InterestProfile)):
+        self._validate_return_type(component_spec, output)
+        state._last = component_spec.output
+        state[state._last] = output
+
+        return state
+
+    def _validate_input_types(self, spec: ComponentSpec, state_types: dict[str, type]):
+        # Best effort type-checking to highlight when inputs are mismatched or out of order
+        sig = signature(spec.component.__call__)
+        sig_params = list(sig.parameters.values())
+
+        for input_name, sig_param in zip(spec.inputs, sig_params):
+            input_type = state_types[input_name]
+            sig_param_type = sig_param.annotation
+            if not included_in(input_type, sig_param_type):
+                msg = (
+                    f"Component {spec.component} expected inputs with types {[p.annotation for p in sig_params]} "
+                    f"but received inputs with types {[state_types[i] for i in spec.inputs]}"
+                )
+
+                raise TypeError(msg)
+
+    def _validate_output_type(self, spec: ComponentSpec):
+        # Best effort type-checking to highlight when output is
+        # mismatched with existing state values types
+        sig = signature(spec.component.__call__)
+        output_name = spec.output
+
+        state_type = self._state_types.get(output_name, None)
+        output_type = sig.return_annotation
+
+        if output_type is not _empty and output_type not in (ArticleSet, InterestProfile):
```

Comment:
would use `issubclass` instead of `not in`

Reply:
Same question here: do we expect to subclass those and if so why?